neqo-transport/src/recovery.rs

```diff
@@ -543,6 +560,53 @@ impl LossRecovery {
             .min_by_key(|&(_, time)| time)
             .map(|(spc, val)| (spc, val + self.loss_delay()))
     }
+
+    /// Find when the last sent ack eliciting packete.
+    pub fn get_min_pto(&self) -> Option<Instant> {
+        // TODO ignore PNSpace::Application until handshake is done -> a server side problem.
+        PNSpace::iter()
+            .filter_map(|spc| {
+                if self.spaces[*spc]
+                    .sent_packets
+                    .values()
+                    .any(|sp| sp.ack_eliciting)
+                {
+                    // None is ordered less than Some(_). Bad. Filter them out.
+                    if let Some(time) = self.spaces[*spc].time_of_last_sent_ack_eliciting_packet {
+                        Some(time + self.pto_timeout(*spc))
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
+            })
+            .min_by_key(|&time| time)
+    }
+
+    pub fn get_min_pto_pn_space(&self, now: Instant) -> Option<PNSpace> {
+        PNSpace::iter()
+            .filter_map(|spc| {
+                if self.spaces[*spc]
+                    .sent_packets
+                    .values()
+                    .any(|sp| sp.ack_eliciting)
+                {
+                    if let Some(time) = self.spaces[*spc].time_of_last_sent_ack_eliciting_packet {
+                        if time <= (now - self.pto_timeout(*spc)) {
+                            Some(*spc)
+                        } else {
+                            None
+                        }
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
+            })
+            .min_by_key(|&spc| spc)
```

Comment:
This might read better if you factor out the function that produces the iterator, so that these functions only do `min_by_key()` with different keys.

Reply:
I can see the difference now. This is good, but I would prefer a little cleaner factoring.